tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +124,125 @@ message WriteScalarResponse {
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The first request should send no actual data; it acts as a request for the
+//    current status of the blob, in case it has already been partially or fully
+//    uploaded.
+//  * The first WriteBlobResponse reports the state, size, and crc32c of the
+//    object so far, i.e. taking into account any previously uploaded data.
+//  * The client proceeds to upload from the specified offset (optionally
+//    validating the crc32c up to that point first).
+//  * The caller sends a number of additional WriteBlobRequests, providing the
+//    blob data in an ordered sequence of chunks.
+//  * The caller receives a sequence of WriteBlobResponses in return.  Each
+//    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
+//    the state of the write operation.
+//  * In general, the caller should not wait to verify that each response
+//    reports the expected status, size, and crc32c before sending the next
+//    request.  Doing so harms throughput, with little benefit since any
+//    mismatch on the server side will cause the RPC to throw an error anyway.
+//  * The request size should be less than 4 MiB (2^20 = 4,194,304), because
+//    that is the default request limit for gRPC. Accounting for the other
+//    fields of the request, then, and leaving some padding for safety, the data
+//    chunk should be somewhat smaller-- perhaps simply 4e6 bytes.
+//  * The last of these requests must set finalize_object=True, and the prior
+//    ones must not.
+//  * For a blob <= 4MB, a single data chunk suffices, so there are two
+//    requests: one to learn the status, and one to upload the data.
+//
+// Note that the WriteBlobRequest does not mirror the nested structure of
+// WriteScalarRequest, because we only ever send one blob at a time.
+
+message CreateBlobSequenceRequest {
+  string experiment_id = 1;
+  string run = 2;
+  string tag = 3;
+  int64 step = 4;
+  int64 final_sequence_length = 5;
+  // Note that metadata.plugin_data.content does not carry the payload.
+  .tensorboard.SummaryMetadata metadata = 6;
+}
+
+message CreateBlobSequenceResponse {
+  string sequence_id = 1;
+}
+
+// Metadata included with only the first chunk of the blob write stream.
+message WriteBlobFirstRequestData {
+  string sequence_id = 1;
+  int64 index = 2;
+}
+
+// A single chunk of the blob write stream.
+message WriteBlobRequest {
+  WriteBlobFirstRequestData write_blob_first_request_data = 1;
+  // The bytes in this chunk.  For the first chunk only, this value should be
+  // empty.  This allows determining from the response how much of the blob has
+  // already been uploaded, if any, without sending redundant data.
+  bytes data = 2 [ctype = CORD];
+  // The position in the blob where this chunk begins.
+  // This must equal the sum of the sizes of the chunks sent so far (except on
+  // the first request).
```

Comment:
Clearer I think to state that this field is ignored if `data` is empty.  (That should be true regardless of whether we preserve the special first request semantics, and is also less ambiguous about whether the validation is based on the actual arrival sequence of the request at the server, or on the contents of the rest of the request message.)

Reply:
+1, thanks